erigon-lib/state/aggregator_bench_test.go

```diff
@@ -287,3 +288,91 @@ func Benchmark_Recsplit_Find_ExternalFile(b *testing.B) {
 		require.EqualValues(b, keys[p], key)
 	}
 }
+
+func BenchmarkAggregator_BeginFilesRo_Latency(b *testing.B) {
+	//BenchmarkAggregator_BeginFilesRo/begin_files_ro-16  1737404  737.3 ns/op  3216 B/op  21 allocs/op
+	aggStep := uint64(100_00)
+	_, agg := testDbAndAggregatorBench(b, aggStep)
+
+	b.Run("begin_files_ro", func(b *testing.B) {
+		for i := 0; i < b.N; i++ {
+			agg.BeginFilesRo()
+		}
+	})
+}
+
+var parallel = flag.Int("bench.parallel", 1, "parallelism value") // runs 1 *maxprocs
+var loopv = flag.Int("bench.loopv", 100000, "loop value")
+
+func BenchmarkAggregator_BeginFilesRo_Throughput(b *testing.B) {
+	/**
+	for cpu in $(seq 0 20); do
+		cpus=$((1 << $cpu))  # Same as 2^cpu
+		echo -n "($cpus, "
+		echo -n $(go test -benchmem -run=^$ -bench ^BenchmarkAggregator_BeginFilesRo_Throughput$ github.com/erigontech/erigon-lib/state  \
+		-bench.parallel=$cpus -bench.loopv=1000 | grep 'BenchmarkAggregator_BeginFilesRo_Throughput' | cut -f3 | xargs|cut -d' ' -f1)
+		echo -n "), "
+	done
+	**/
+	// trying to find BeginFilesRo throughput
+	// result: deteriorates after 2^21 goroutines.
+	if !flag.Parsed() {
+		flag.Parse()
+	}
+	//b.Logf("Running with parallel=%d work=%d, #goroutines:%d", *parallel, *loopv, *parallel*runtime.GOMAXPROCS(0))
+
+	aggStep := uint64(100_00)
+	_, agg := testDbAndAggregatorBench(b, aggStep)
+
+	b.SetParallelism(*parallel) // p * maxprocs
+	b.RunParallel(func(b *testing.PB) {
+		foo := 0
+		for b.Next() {
+			tx := agg.BeginFilesRo()
+			for i := 0; i < *loopv; i++ {
+				foo *= 2
+				foo /= 2
+			}
+			tx.Close()
+		}
+	})
+}
+
+func BenchmarkDb_BeginFiles_Throughput(b *testing.B) {
+	/**
+	for cpu in $(seq 0 20); do
+	    cpus=$((1 << $cpu))  # Same as 2^cpu
+	    echo -n "($cpus, "
+	    echo -n $(go test -benchmem -run=^$ -bench ^BenchmarkDb_BeginFiles_Throughput$ github.com/erigontech/erigon-lib/state  \
+		-bench.parallel=$cpus -bench.loopv=1000 | grep 'BenchmarkDb_BeginFiles_Throughput' | cut -f3 | xargs|cut -d' ' -f1)
+	    echo -n "), "
+	done
+	**/
+
+	// trying to find BeginFilesRo and Rollback throughput
+	// result: deteriorates after 2^21 goroutines.
+	if !flag.Parsed() {
+		flag.Parse()
+	}
+	//b.Logf("Running with parallel=%d work=%d, #goroutines:%d", *parallel, *loopv, *parallel*runtime.GOMAXPROCS(0))
+
+	aggStep := uint64(100_00)
+	db, _ := testDbAndAggregatorBench(b, aggStep)
+	ctx := context.Background()
+
+	b.SetParallelism(*parallel) // p * maxprocs
+	b.RunParallel(func(pb *testing.PB) {
+		foo := 0
+		for pb.Next() {
+			tx, err := db.BeginRo(ctx)
+			if err != nil {
+				b.Fatalf("%v", err)
+			}
+			for i := 0; i < *loopv; i++ {
```

Comment:
Maybe CPU become a bottleneck. try `time.Sleep` as emulation of IO-pageFault.

Reply:
added a benchmark with time.Sleep 
- deterioration after 130k gorouines.